pkg/shared/callback/callback.go

```diff
@@ -0,0 +1,285 @@
+		// read message.
+		callbackTime := time.Now().UnixMilli()
+		newObject := Request{
+			Vertex:     u.vertexName,
+			Pipeline:   u.pipelineName,
+			ID:         uuid,
+			CbTime:     callbackTime,
+			FromVertex: msg.ID.VertexName,
+		}
+
+		// if the callback URL is not present in the map, create a new slice
+		if _, ok = callbackUrlMap[callbackURL]; !ok {
+			callbackUrlMap[callbackURL] = make([]Request, 0)
+		}
+
+		// Add new CallbackResponse to map, grouped by the Callback URL
+		callbackUrlMap[callbackURL] = append(callbackUrlMap[callbackURL], newObject)
+	}
+
+	if err := u.executeCallback(ctx, callbackUrlMap); err != nil {
+		return fmt.Errorf("error executing callback: %w", err)
+	}
+
+	return nil
+}
+
+// executeCallback sends POST requests to the provided callback URLs with the corresponding request bodies.
+// It returns an error if any of the requests fail.
+func (u *Uploader) executeCallback(ctx context.Context, callbackUrlMap map[string][]Request) error {
+	var failedRequests []Request
+
+	for url, requests := range callbackUrlMap {
+		err := u.sendRequest(ctx, url, requests)
+		if err != nil {
+			u.opts.logger.Errorw("Failed to send request, will try writing to the callback URL",
+				zap.String("url", url),
+				zap.Error(err),
+			)
+			failedRequests = append(failedRequests, requests...)
+		}
+	}
+
+	if len(failedRequests) > 0 && u.opts.callbackURL != "" {
+		err := u.sendRequest(ctx, u.opts.callbackURL, failedRequests)
+		if err != nil {
+			u.opts.logger.Errorw("Failed to send request to callback URL, skipping the callback requests",
+				zap.String("url", u.opts.callbackURL),
+				zap.Error(err),
+			)
+		}
+	}
+
+	return nil
+}
+
+// sendRequest sends a POST request to the provided URL with the provided requests.
+// It returns an error if the request fails or if the response status is not OK.
+func (u *Uploader) sendRequest(ctx context.Context, url string, requests []Request) error {
+	// get a cached client
+	client := u.GetClient(url)
+
+	body, err := json.Marshal(requests)
+	if err != nil {
+		return fmt.Errorf("failed to marshal requests: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode > 299 {
+		return fmt.Errorf("received non-OK response status: %s", resp.Status)
+	}
+
+	return nil
+}
+
+// GetClient returns a client for the given URL from the cache
+// If the client is not in the cache, a new one is created.
+func (u *Uploader) GetClient(url string) *http.Client {
+	// Check if the client is in the cache
+	if client, ok := u.clientsCache.Get(url); ok {
+		return client
+	}
+
+	// If the client is not in the cache, create a new one
+	client := &http.Client{
```

Comment:
if the callback server (for this url) is restarted, will the http.Client auto-reconnect internally or will this client become invalid?

Reply:
For the server restart case, If I am not wrong the client doesn't maintain a persistent connection to the server, so it simply creates a new connection when there is a post request, so ideally it should be fine. But still, the client caches the TCP connections. Let me run some tests, If I find any issues I will create a follow-up PR to fix it.